### PR TITLE
Dm 3533

### PR DESCRIPTION
**Pull request by JoshingYou1**
### JIRA issue link
https://agile6.atlassian.net/browse/DM-3533

## Description - what does this code do?
Adds a new `CompoundBodyComponent` to page-builder that allows the user to add text and images.

## Testing done - how did you test it/steps on how can another person can test it 
_**Prerequisite:**_ In the terminal, run `rails db:migrate`.
1. Log in as an admin and visit `/admin/pages`.
2. From here, you can either choose to create a new page or edit an existing page.
3. Once on the create/edit page, click on the "Add New page component" button and select the "Text and Images" option.
4. **IMPORTANT:** No attributes for the `CompoundBodyComponent`, so in theory you could only have images, but it's not recommended.
5. Fill out any fields you like, and once you're done, click on the "Add image" below the "Text alignment" field.
6. This should reveal an "Image list" banner along with the fields for a single `PageComponentImage`.
7. **IMPORTANT:** Unlike the `CompoundBodyComponent`, there are two required fields for a `PageComponentImage`, `image` (should be a given lolz) and `alt_text`.
8. Fill out the all of the fields for this component image (add some links and cool stuff to the caption to spice it up 🌶️ ).
9. Take a minute to review what you did so you have an idea in your head of how the component should behave on the Page's show page.
10. Go ahead and click "Update/Create page", and make sure it saved properly (on the admin show page of the Page, you should see all of the attributes you selected for both the `CompoundBodyComponent` and the `PageComponentImage`).
11. If you're on the admin panel show page of the Page still, just click on the "COMPLETE URL" link that way it'll open up a new browser window so you can compare the data on the show page and the admin show page. Also make sure that the styles match the Figma designs.
12. If everything looks good, visit `/admin/pages/#{your page's id}/edit` to get back to the edit view, or just click the "Edit Page" button at the top right of the admin show page.
13. Here's where I want you to play around with the images a bit. Add some, delete some, leave required fields blank, edit some fields, etc. The backend should be set up in a way that, if there are visible `PageImageComponents` (ones not marked to be deleted) that have missing required fields (`image` upload, `alt_text`), they will be removed from the params, BUT all of the other complete images and the other parts of the Page should save correctly. In this instance, you should be redirected to the admin show page for the Page, and a message stating: "One or more 'Compound Body' components had missing required fields for its image(s). The page was saved, but those image(s) were not." should pop up.
14. If you got the warning message from the last part of step 13, click "Edit Page". again and make sure the images you DID complete or edit (but still had the required fields completed for) are still in tact and the values reflect your changes.
15. Refresh the Page's show page. The next steps will go in depth about the styles for the different scenarios that can surface.
### Scenario 1: ONLY text
- Based on the `text_alignment` value, the text should be displayed on that right side of the page (left or right).
- The text should take up six columns.
- If the "LARGER FONT SIZE" checkbox was selected, the title's `h2` element should have the `usa-prose-h1` class, which increases the font size by a bit (compared to `usa-prose-h2`).
- From here, make sure the other values you selected are correctly reflected (margin bottom, margin top, URL, etc.).
### Scenario 2: Text and a single image
- Based on the `text_alignment` value, the text should be displayed on that side of the page (left or right) and the image should be displayed on the opposite site.
- The text should take up 5 columns and the image should take up 4 columns (based on Figma designs).
- For the image, if you added a URL, make sure the image is clickable and the URL matches the value you entered.
- For the image, if you added a caption, it should be placed below the image
### Scenario 3: Text and a multiple images (no designs for this scenario, so these styles may change in the future)
- The text styles should not change from scenario 2.
- For the images, the container that holds them is its own grid system, and the images should be spread amongst the columns as evenly as possible (Ex: 2 images take up 6 columns each, 3 images take up 4 columns each, etc.)
### Scenario 4: No text (`title` and `text` are not present) and image(s) (no designs for this scenario, so these styles may change in the future)
- Basically, the images will behave the same as if there was text (on the opposite side of the `text_alignment` value). In other words, you can move the images from side to side, but they will still only be on one side of the page.
- The images also behave the same way as scenario 3.

## Mobile (minimum `340px`)
- All text and images should take up 12 columns (regardless of how many images are present for a given `CompoundBodyComponent`).
- The images should always sit on top of the text, if text is present. If not, the images will sit on top of each other based on the order they were uploaded on the backend.

### Important final note!!!
- Even though you've selected a `margin_bottom` for a `CompoundBodyComponent`, IF the component itself in the last one for the `Page`, it will default to a margin bottom of `0`.

## Link to mock-ups/mock ups (image file if you have it) (if applicable)
https://www.figma.com/file/NYiAmX7l9yGESH9sTkX7IS/XR-Drafts?node-id=554%3A3546

## Acceptance criteria
- Header above title 
- If field is blank, do not display
- Title (Bolded, larger text)
- Admin can select Big text header or little text header with checkbox
- WYSIWYG editor for body text
- Bolded link (for another page) (e.g. View all Innovations) (URL input)
- Ability to upload image(s)
- optional caption with WYSIWYG
- Image can be link with URL input
- If section has image, make it 50/50% of page (e.g. Featured Innovation)
- If no imagine, make it go a little longer
- Option for which side the text is on (alignment? side?)
- AS to confirm: If user chooses 50/50 split with image, image is required

## Definition of done
- [X] Unit tests written (if applicable)
- [X] e2e/accessibility tests written (if applicable)
- [X] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [X] A link has been provided to the originating JIRA issue
- [X] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

### CODE REVIEW

File: app/views/active_admin/resource/_page_compound_body_component_form.html.arb [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+               class: 'inline-hints'
+        end
+
+        # Page component images header bar
+        div class: "page-component-images-header display-flex flex-align-center #{'display-none' if page_component_images.blank?}" do
+          span 'Image list'
+        end
+
+        # PageComponentImage list
+        ol id: "page_page_components_attributes_#{placeholder}_page_component_images_ol", class: 'page-component-images-ol' do
```
Comment (JoshingYou1) [inline app/views/active_admin/resource/_page_compound_body_component_form.html.arb:154]: One thing I was unable to figure out how to do was to make the `PageComponentImage` list a partial so we could use it for other components in the future. If you have any examples of how to do this, feel free to let me know and I can make the necessary changes.

**Review by bradjohnson92008 — COMMENTED** [review]
Display and content look to be working as described in the PR.  Only thing I really noticed was that external urls created in the WYSIWYG are not displaying as external urls - i thought we had that baked in to the WYSIWYG component?  I need to do some additional testing tomorrow but looking great so far!  Great work Josh!

**Review by bradjohnson92008 — APPROVED** [review]
Looks Good Josh!  You've made a really flexible component here nice work!  Regarding the imagelist component as partial... that is probably doable but lets keep that as a V2 thing.. unless you can see an easy way to refactor it.. and we have time...  lol.

File: app/helpers/application_helper.rb [hunk 2]
```diff
@@ -125,8 +125,14 @@ def get_grid_alignment_css_class(alignment)
     end
   end
 
+  def is_internal_link?(url)
```
Comment (camillevilla) [inline app/helpers/application_helper.rb:128]: LOL I ended up making this method in my PR too 😂  well done

**Review by camillevilla — CHANGES_REQUESTED** [review]
This looking great, @JoshingYou1! I definitely want to back to News and Events and add some validations like what you've got in your models. 

One small change request: can you remove external link styling from the images?
Screen Shot 2022-09-28 at 3 25 15 PM

Here's[ some USWDS guidance](https://designsystem.digital.gov/components/link/):
>If you use an external link indicator, use it consistently for all text links. If your project uses an external link indicator (like an icon), use it for all text links across your site. If users learn to associate an external link with the indicator, they will also expect that text links without an indicator are not external links. Icon- or image-only links like social share buttons or logos do not need the same treatment as text links.

File: app/assets/javascripts/_page_show.es6 [hunk 3]
```diff
@@ -1,3 +1,9 @@
+const COMPONENT_CLASSES = [
```
Comment (camillevilla) [inline app/assets/javascripts/_page_show.es6:1]: nice refactor! This looks neat. 

For removing the images from external link styling, you could do something like `.page-compound-body-component p` to get the WYSIWIG paragraphs. For the featured link, you could throw a class on the `a` tag and then add something like `.page-compound-body-component .featured-url` to this list

Comment (JoshingYou1) [reply app/assets/javascripts/_page_show.es6:1]: > nice refactor! This looks neat.
> 
> For removing the images from external link styling, you could do something like `.page-compound-body-component p` to get the WYSIWIG paragraphs. For the featured link, you could throw a class on the `a` tag and then add something like `.page-compound-body-component .featured-url` to this list

This issue has been resolved! Nice catch 😄

**Review by camillevilla — APPROVED** [review]

**Pull request merged**